File: trunk/SDKs/Java/src/com/tokbox/api/BaseAPI.java
/*
*		*************************************
*		*       TokBox Base API             *
*		*************************************
*
*		Melih Onvural, August 2008
*
*
*/

package com.tokbox.api;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.UnsupportedEncodingException;
import java.math.BigInteger;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLEncoder;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.Map;
import java.util.Random;
import java.util.logging.Logger;

import com.tokbox.exception.TokBoxRuntimeException;
import com.tokbox.util.TokBoxUtils;

public class BaseAPI {
	private String partnerKey;
	private String secret;
	
	private String jabberId;

	private Logger logger; 

	private final String version = "1.0.0";
	private final int sigMethod = BaseAPI.SIMPLE_MD5;
	
	public final static int SIMPLE_MD5 = 1;
	public final static int HMAC_SHA1 = 2;
	
	public final static String API_SERVER_LOGIN_URL = "view/oauth&";
	public final static String API_SERVER_METHODS_URL = "a/v0";
	public final static String API_SERVER_CALL_WIDGET = "vc/";
	public final static String API_SERVER_RECORDER_WIDGET = "vr/";
	public final static String API_SERVER_PLAYER_WIDGET = "vp/";
	
	public BaseAPI(String partnerKey, String secret) {
		this.partnerKey = partnerKey;
		this.secret = secret;
		this.jabberId = "";
		
		if(API_Config.DEBUG_MODE) {
			this.logger = Logger.getLogger("com.tokbox.api.BaseAPI");
		}
	}

	public String getPartnerKey() {
		return this.partnerKey;
	}

	public String getSecret() {
		return this.secret;
	}

	public String getJabberId() {
		return this.jabberId;
	}

	public void setJabberId(String jabberId) {
		this.jabberId = jabberId;
	}

	public void setSecret(String secret) {
		this.secret = secret;
	}

	public String sigMethodText() {
		return this.sigMethod == BaseAPI.SIMPLE_MD5 ? "SIMPLE-MD5" : "HMAC-SHA1";
	}

	protected String TB_Request(String method, String apiUrl, Map<String, String> paramList) {
		StringBuilder returnString = new StringBuilder();

		URL url = null;
		HttpURLConnection conn = null;
		BufferedReader br = null;
		OutputStreamWriter wr = null;
		BufferedWriter bufWriter = null;
		try {
			String reqString = API_Config.API_SERVER+BaseAPI.API_SERVER_METHODS_URL+apiUrl;
			String nonce = this.generateNonce();
			String timestamp = Long.toString(System.currentTimeMillis());

			StringBuilder dataString = new StringBuilder();

			for(Iterator<String> i = paramList.keySet().iterator(); i.hasNext(); ) {
				String key = i.next();
				String value = paramList.get(key);

				if(null != value) {
					value = URLEncoder.encode(paramList.get(key), "UTF-8").replaceAll("\\+", "%20");
					dataString.append(URLEncoder.encode(key, "UTF-8")).append("=").append(value).append("&");
				}
			}

			String signedSig = this.buildSignedRequest(method, reqString, nonce, timestamp, paramList);

			//Add oauth_version to parameter string
			dataString.append("oauth_version=").append(this.version).append("&");

			//Add timestamp to parameter string
			dataString.append("oauth_timestamp=").append(timestamp).append("&");

			//Add nonce to parameter string
			dataString.append("oauth_nonce=").append(nonce).append("&");

			//Add tokbox_jabber_id to parameter string
			dataString.append("tokbox_jabberid=").append(URLEncoder.encode(this.jabberId, "UTF-8")).append("&");

			//Add oauth_partner_key to parameter string
			dataString.append("oauth_partner_key=").append(this.partnerKey).append("&");

			//Add oauth_signature_method to parameter string
			dataString.append("oauth_signature_method=").append(this.sigMethodText()).append("&");

			//Add the _AUTHORIZATION to parameter string
			dataString.append("_AUTHORIZATION=");

			//Adding the oauth_partner_key to the _AUTHORIZATION 
			dataString.append("oauth_partner_key=\"").append(URLEncoder.encode(this.partnerKey, "UTF-8")).append("\",");

			//Adding the oauth_signature_method to the _AUTHORIZATION
			dataString.append("oauth_signature_method=\"").append(URLEncoder.encode(this.sigMethodText(), "UTF-8")).append("\",");

			//Adding the oauth_timestamp to the _AUTHORIZATION
			dataString.append("oauth_timestamp=\"").append(URLEncoder.encode(timestamp, "UTF-8")).append("\",");

			//Adding oauth_version to the _AUTHORIZATION
			dataString.append("oauth_version=\"").append(URLEncoder.encode(this.version, "UTF-8")).append("\",");

			//Adding oauth_nonce to the _AUTHORIZATION
			dataString.append("oauth_nonce=\"").append(URLEncoder.encode(nonce, "UTF-8")).append("\",");

			//Adding oauth_signature to the _AUTHORIZATION
			dataString.append("oauth_signature=\"").append(URLEncoder.encode(signedSig, "UTF-8")).append("\",");

			//Adding tokbox_jabberid to the _AUTHORIZATION
			dataString.append("tokbox_jabberid=\"").append(URLEncoder.encode(this.jabberId, "UTF-8")).append("\"");

			if(API_Config.DEBUG_MODE) {
				logger.info(dataString.toString());
			}

			url = new URL(reqString);
			conn = (HttpURLConnection) url.openConnection();

			conn.setDoOutput(true);
			conn.setDoInput(true);
			conn.setUseCaches(false);

			conn.setRequestMethod("GET");
			conn.setRequestProperty("Content-Type", "application/x-www-form-urlencoded; charset=UTF-8");
			conn.setRequestProperty("Content-Length", Integer.toString(dataString.toString().length()));
			conn.setRequestProperty("Accept-Charset", "utf-8");
			conn.setRequestProperty("Accept", "text/html, application/xhtml+xml,application/xml");

			wr = new OutputStreamWriter(conn.getOutputStream(), "UTF8");
            bufWriter = new BufferedWriter( wr );
            bufWriter.write(dataString.toString());
            bufWriter.flush();

            br = new BufferedReader(new InputStreamReader(conn.getInputStream(), "UTF8"));

			String str;
			while(null != ((str = br.readLine())))
			{
				returnString.append(str);
				returnString.append("\n");
			}
			
			if(API_Config.DEBUG_MODE) {
				logger.info(returnString.toString());
			}
		} catch(IOException e) {
			e.printStackTrace();
		} finally {
			try {
				if(null != conn) {
					conn.disconnect();
				}
				
				if(null != wr) {
					wr.close();
				}
				
				if(null != bufWriter) {
					bufWriter.close();
				}
				
				if(null != br) {
					br.close();
				}
			}
			catch(IOException e) {
				e.printStackTrace();
			}
		}
		
		return returnString.toString();
	}

	private String buildSignedRequest(String method, String uri, String nonce, String timestamp, Map<String, String> paramList) throws UnsupportedEncodingException {
		String signedString = null;

		paramList.put("oauth_partner_key", this.partnerKey);
		paramList.put("oauth_signature_method", this.sigMethodText());
		paramList.put("oauth_timestamp", timestamp);
		paramList.put("oauth_version", this.version);
		paramList.put("oauth_nonce", nonce);
		paramList.put("tokbox_jabberid", this.jabberId);

		StringBuilder requestString = new StringBuilder();
		requestString.append(method).append("&").append(uri).append("&").append(this.generateRequestString(paramList)).append(this.secret);

		switch(this.sigMethod) {
		case BaseAPI.SIMPLE_MD5:
			signedString = this.md5Hash(requestString.toString());
			break;
		case BaseAPI.HMAC_SHA1:
			//not currently implemented
			break;
		default:
			throw new TokBoxRuntimeException("Desired Signature Method is not supported", new Throwable());
		}

		if(API_Config.DEBUG_MODE) {
			logger.info(requestString.toString());
			logger.info(signedString);
		}

		return signedString;
	}

	private String generateRequestString(Map<String, String> paramList) throws UnsupportedEncodingException {
		ArrayList<String> encodedParamList = new ArrayList<String>();

		for(Iterator<String> i = paramList.keySet().iterator(); i.hasNext(); ) {
			String key = i.next();
			
			if(key.charAt(0) == '_') {
				continue;
			}
			
			String value = paramList.get(key);
			
			if((null != value) && (value.length() > 0)) {
				String encodedKey = URLEncoder.encode(key, "UTF-8");
				String encodedValue = URLEncoder.encode(value, "UTF-8").replaceAll("\\+", "%20");

				StringBuilder encodedString = new StringBuilder();
				encodedString.append(encodedKey).append("=").append(encodedValue);

				encodedParamList.add(encodedString.toString());
			}
		}
		
		Collections.sort(encodedParamList);
		
		return TokBoxUtils.join(encodedParamList, "&");
	}
	
	private String generateNonce() {
		StringBuilder nonce = new StringBuilder(16);
		int length = 16;
		Random generator = new Random();
		
		String seed = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
		int seedLength = seed.length() - 1;
		
		for(int i = 0; i < length; i++) {
			nonce.append(seed.charAt(generator.nextInt(seedLength)));
		}
		
		return nonce.toString();
	}
	
	private String md5Hash(String token) {
		String hashword = null;
		try {
			MessageDigest md5 = MessageDigest.getInstance("MD5");
			md5.update(token.getBytes());
			BigInteger hash = new BigInteger(1, md5.digest());
			hashword = hash.toString(16);
		} catch (NoSuchAlgorithmException nsae) {
			
		}
		
		return pad(hashword, 32, '0');
	}

	private String pad(String s, int length, char pad) {
		StringBuffer buffer = new StringBuffer(s);
		while (buffer.length() < length) {
			buffer.insert(0, pad);
		}

		return buffer.toString();
	}
}

File: SDKs/dotNET/BaseAPI.cs
/*
 * Copyright (c) 2008 Syedur Islam
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *		
 *************************************

 *
 *       TokBox Base API             *
 *
 *************************************
 *
 * Original Java Code by Melih Onvural, August 2008

 * Converted to C# by Syedur Islam, November 2008

 *

 *

 */

using System;
using System.Collections;
using System.Collections.Generic;
using System.Net;
using System.Text;
using System.IO;
using System.Security.Cryptography;
using System.Web;

namespace TokBox
{
    public class BaseAPI
    {
        private string partnerKey;
        private string secret;

        private string jabberId;

        private const string version = "1.0.0";
        private const int sigMethod = SIMPLE_MD5;

        public const int SIMPLE_MD5 = 1;
        public const int HMAC_SHA1 = 2;

        public const string API_SERVER_LOGIN_URL = "view/oauth&";
        public const string API_SERVER_METHODS_URL = "a/v0";
        public const string API_SERVER_CALL_WIDGET = "vc/";
        public const string API_SERVER_RECORDER_WIDGET = "vr/";
        public const string API_SERVER_PLAYER_WIDGET = "vp/";

        public BaseAPI(string partnerKey, string secret)
        {
            this.partnerKey = partnerKey;
            this.secret = secret;
            this.jabberId = "";
        }

        public string GetPartnerKey()
        {
            return this.partnerKey;
        }

        public string GetSecret()
        {
            return this.secret;
        }

        public string GetJabberId()
        {
            return this.jabberId;
        }

        public void SetJabberId(string jabberId)
        {
            this.jabberId = jabberId;
        }

        public void SetSecret(string secret)
        {
            this.secret = secret;
        }

        public string SigMethodText()
        {
            return sigMethod == SIMPLE_MD5 ? "SIMPLE-MD5" : "HMAC-SHA1";
        }

        protected string TB_Request(string method, string apiUrl, Dictionary<string, string> paramList)
        {
            string reqString = API_Config.API_SERVER + API_SERVER_METHODS_URL + apiUrl;

            WebRequest request = WebRequest.Create(reqString);
            try
            {
                string nonce = this.GenerateNonce();
                
				//string timestamp = DateTime.Now.Millisecond.ToString();
				//generate a unix timestamp -- modified by Volkan Özçelik @2009-01-09,21:36,GMT+2
				TimeSpan span = DateTime.Now - (new DateTime(1970, 1, 1, 0, 0, 0, 0));
				string timestamp = "" + Convert.ToInt64(span.TotalSeconds);

				string signedSig = this.BuildSignedRequest(method, reqString, nonce, timestamp, paramList);

				StringBuilder dataString = new StringBuilder();

                foreach (KeyValuePair<string, string> kvp in paramList)
                {
                    dataString.Append(HttpUtility.UrlEncode(kvp.Key, Encoding.UTF8));
                    dataString.Append("=");
                    dataString.Append(HttpUtility.UrlEncode(kvp.Value, Encoding.UTF8).Replace("+", "%20"));
                    dataString.Append("&");
                }


                //Add the _AUTHORIZATION to parameter string
                dataString.Append("_AUTHORIZATION=");

                //Adding the oauth_partner_key to the _AUTHORIZATION 
                dataString.Append("oauth_partner_key=\"").Append(HttpUtility.UrlEncode(partnerKey, Encoding.UTF8)).Append("\",");

                //Adding the oauth_signature_method to the _AUTHORIZATION
                dataString.Append("oauth_signature_method=\"").Append(HttpUtility.UrlEncode(SigMethodText(), Encoding.UTF8)).Append("\",");

                //Adding the oauth_timestamp to the _AUTHORIZATION
                dataString.Append("oauth_timestamp=\"").Append(HttpUtility.UrlEncode(timestamp, Encoding.UTF8)).Append("\",");

                //Adding oauth_version to the _AUTHORIZATION
                dataString.Append("oauth_version=\"").Append(HttpUtility.UrlEncode(version, Encoding.UTF8)).Append("\",");

                //Adding oauth_nonce to the _AUTHORIZATION
                dataString.Append("oauth_nonce=\"").Append(HttpUtility.UrlEncode(nonce, Encoding.UTF8)).Append("\",");

                //Adding oauth_signature to the _AUTHORIZATION
                dataString.Append("oauth_signature=\"").Append(HttpUtility.UrlEncode(signedSig, Encoding.UTF8)).Append("\",");

                //Adding tokbox_jabberid to the _AUTHORIZATION
                dataString.Append("tokbox_jabberid=\"").Append(HttpUtility.UrlEncode(this.jabberId, Encoding.UTF8)).Append("\"");

                byte[] bytes = Encoding.UTF8.GetBytes(dataString.ToString());

                request.ContentType = "application/x-www-form-urlencoded";
                request.ContentLength = bytes.Length;
                request.Method = method;

                using (Stream requestStream = request.GetRequestStream())
                {
                    requestStream.Write(bytes, 0, bytes.Length);

                    try
                    {
                        using (WebResponse response = request.GetResponse())
                        {
                            using (StreamReader reader = new StreamReader(response.GetResponseStream()))
                            {
                                return reader.ReadToEnd();
                            }
                        }
                    }
                    catch (Exception ex)
                    {
                        throw new Exception(ex.StackTrace);
                    }
                }
            }
            catch (Exception ex)
            {
                throw new Exception(ex.StackTrace);
            }

        }

        private string BuildSignedRequest(string method, string uri, string nonce, string timestamp, Dictionary<string, string> paramList)
        {
            string signedString = null;

            paramList.Add("oauth_partner_key", this.partnerKey);
            paramList.Add("oauth_signature_method", this.SigMethodText());
            paramList.Add("oauth_timestamp", timestamp);
            paramList.Add("oauth_version", version);
            paramList.Add("oauth_nonce", nonce);
            paramList.Add("tokbox_jabberid", this.jabberId);

            StringBuilder requestString = new StringBuilder();
            requestString.Append(method).Append("&").Append(uri).Append("&").Append(GenerateRequestString(paramList)).Append(secret);

            Console.WriteLine("Request string is: " + requestString);

            switch (sigMethod)
            {
                case SIMPLE_MD5:
                    signedString = MD5Hash(requestString.ToString());
                    break;
                case HMAC_SHA1:
                    //not currently implemented
                    break;
                default:
                    throw new TokBoxRuntimeException("Desired Signature Method is not supported", null);
            }

            Console.WriteLine("Signed string is: " + signedString);

            return signedString;
        }

        private string GenerateRequestString(Dictionary<string, string> paramList)
        {
            List<string> encodedParamList = new List<string>();

            foreach (KeyValuePair<string, string> kvp in paramList)
            {
                if (!string.IsNullOrEmpty(kvp.Key) && !string.IsNullOrEmpty(kvp.Value))
                {
                    StringBuilder encodedString = new StringBuilder();

                    string encodedKey = HttpUtility.UrlEncode(kvp.Key, Encoding.UTF8);
                    string encodedValue = HttpUtility.UrlEncode(kvp.Value, Encoding.UTF8)
                        .Replace("+", "%20")
                        .Replace("%3a", "%3A")
                        .Replace("%2f","%2F");
                    encodedString.AppendFormat("{0}={1}", encodedKey, encodedValue);

                    encodedParamList.Add(encodedString.ToString());
                }
            }

            encodedParamList.Sort();

            Console.WriteLine("EncodedParamList: " + TokBoxUtils.Join(encodedParamList, "&"));
            return TokBoxUtils.Join(encodedParamList, "&");
        }

        private string GenerateNonce()
        {
            StringBuilder nonce = new StringBuilder(16);
            const int length = 16;
            Random generator = new Random();

            const string seed = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
            int seedLength = seed.Length - 1;

            for (int i = 0; i < length; i++)
            {
                nonce.Append(seed[generator.Next(seedLength)]);
            }

            Console.WriteLine("Nonce: " + nonce);
            return nonce.ToString();
        }

        private string MD5Hash(string token)
        {
            MD5 md5 = MD5.Create();
            byte[] data = md5.ComputeHash(Encoding.Default.GetBytes(token));
            StringBuilder sb = new StringBuilder();

            for (int i = 0; i < data.Length; i++)
            {
                sb.Append(data[i].ToString("x2"));
            }

            Console.WriteLine("MD5 Hash: " + sb);
            return sb.ToString();
        }
    }
}